File: ABC410/temp/D_0614.R
options(scipen = 100, digits = 15)
# --- 读取输入 ---
con <- file("stdin", "r")
lines <- readLines(con)
close(con)

tok <- as.integer(unlist(strsplit(lines, "\\s+")))
N <- tok[1]
M <- tok[2]
idx <- 3

# --- 建图（邻接表） ---
G <- vector("list", N)
for (i in seq_len(M)) {
  a <- tok[idx]
  b <- tok[idx + 1]
  w <- tok[idx + 2]
  idx <- idx + 3
  G[[a]] <- append(G[[a]], list(c(b, w)))
}

# --- 初始化 ---
dist <- integer(N) # 灯状态（XOR 值）
vis <- logical(N) # 是否访问
basis <- integer(0) # 工具箱：收集环的 XOR

# --- 插入新钥匙到线性基 ---
insert_basis <- function(x) {
  for (b in basis) {
    x <- min(x, bitwXor(x, b))
  }
  if (x != 0L) {
    basis <<- sort(c(basis, x), decreasing = TRUE)
  }
}

# --- 深度优先搜索 ---
dfs <- function(u) {
  vis[u] <<- TRUE
  for (edge in G[[u]]) {
    v <- edge[1]
    w <- edge[2]
    if (!vis[v]) {
      dist[v] <<- bitwXor(dist[u], w)
      dfs(v)
    } else {
      cycle <- bitwXor(dist[u], bitwXor(w, dist[v]))
      if (cycle != 0L) insert_basis(cycle)
    }
  }
}

# --- 主流程 ---
dfs(1L)

if (!vis[N]) {
  cat(-1L, "\n")
} else {
  res <- dist[N]
  for (b in basis) {
    res <- min(res, bitwXor(res, b))
  }
  cat(res, "\n")
}

File: ABC410/D_0614.py
import sys
sys.setrecursionlimit(10**7)

def main():
    input = sys.stdin.readline
    N, M = map(int, input().split())
    G = [[] for _ in range(N+1)]

    for _ in range(M):
        a, b, w = map(int, input().split())
        G[a].append((b, w))

    dist = [0] * (N + 1)
    vis  = [False] * (N + 1)
    basis = [] 

    def insert_basis(x):
        for b in basis:
            x = min(x, x ^ b)
        if x:
            basis.append(x)
            basis.sort(reverse=True) 
    def dfs(u):
        vis[u] = True
        for v, w in G[u]:
            if not vis[v]:
                dist[v] = dist[u] ^ w
                dfs(v)
            else:
                cycle = dist[u] ^ w ^ dist[v]
                if cycle:
                    insert_basis(cycle)

    dfs(1)

    if not vis[N]:
        print(-1)
        return

    res = dist[N]
    for b in basis:
        res = min(res, res ^ b)

    print(res)

if __name__ == "__main__":
    main()